### ISSUE TITLE
Adding PEC as technique supported by calibration

### ISSUE THREAD

**Issue by nathanshammah** [unitaryfund/mitiq#1737]
## Issue Description
Right now calibration supports ZNE as error mitigation technique. It would be great to also support PEC. 

## Proposed Solution
1. One option would require user input about the noise channel model. This could include a call of:
- mitiq.pec.representations.optimal.**find_optimal_representation**(ideal_operation, noisy_operations)
2. Another option would not require input about noise channel model. These could be functions that could be called: 
- mitiq.pec.representations.depolarizing.**represent_operation_with_global_depolarizing_noise**(ideal_operation, noise_level)
- mitiq.pec.representations.depolarizing.**represent_operation_with_local_depolarizing_noise**(ideal_operation, noise_level)
- mitiq.pec.representations.biased_noise.**represent_operation_with_local_biased_noise**(ideal_operation, epsilon, eta)

## Additional References
https://mitiq.readthedocs.io/en/latest/apidoc.html#mitiq.pec.representations.depolarizing.represent_operation_with_global_depolarizing_noise
https://mitiq.readthedocs.io/en/stable/guide/calibrators.html

**Comment by YuNariai** [comment]
I would like to try to work on this! As I understand, we want to add a PECSettings for PEC, just like ZNESettings for ZNE, right?

**Comment by andreamari** [comment]
Hi @YuNariai, I think defining a simple `PECSettings` dictionary and making sure it works correctly is a good start.

@nathanshammah, suggested more details on the description of this issue. As a first step, when defining PEC strategies, we could start from the second option proposed by @nathanshammah and we could just focus on local depolarizing noise:

`mitiq.pec.representations.depolarizing.represent_operation_with_local_depolarizing_noise(ideal_operation, noise_level)
`
**only applied to two-qubit gates** with different values of `noise_level` e.g. 0.001, 0.005, 0.01, 0.015, 0.02

In this way the `Calibrator` should (hopefully) suggest the value of `noise_level` which gives best performances.

**Comment by andreamari** [comment]
Hi [YuNariai](https://github.com/YuNariai), are you still interested in working on this issue?
Otherwise, no problem, someone else will take it.

**Comment by YuNariai** [comment]
Hi, I'm still working on it, but I may still need some time, if it's urgent, please let someone take it over.

**Comment by andreamari** [comment]
Sounds good! I just assigned the issue to you. For any problem or blocker, let us know. Thanks!

**Comment by YuNariai** [comment]
I'm thinking about using `represent_operations_in_circuit_with_local_depolarizing_noise` to make things easier, do you think I can use it?

**Comment by andreamari** [comment]
> I'm thinking about using `represent_operations_in_circuit_with_local_depolarizing_noise` to make things easier, do you think I can use it?

That function assumes the same amount of noise for all gates. This is not realistic in a real system since two-qubit gates are much more noisy in practice. 

So, it's fine to use `represent_operations_in_circuit_with_local_depolarizing_noise` to get a prototype, but at some point we'll need to define representations for two-qubit gates only, such that PEC is applied only to them.  It's ok to assume that two-qubit gates can only be CNOT or CZ gates, if this restriction helps.

This is an [example](https://mitiq.readthedocs.io/en/stable/examples/pec_tutorial.html) in which PEC is applied to CNOT gates only. 

The recent introduction of the `qubit-independent` [option](https://mitiq.readthedocs.io/en/stable/guide/pec-3-options.html#qubit-independent-representations) for  `OperationRepresentation` is probably useful.

**Comment by YuNariai** [comment]
Ok I got it!
so we want to try 5 noise-level: 0.001, 0.005, 0.01, 0.015, 0.02 on each two-qubit gate, and the cost is 5**(number of two-qubit gates), right?

**Comment by andreamari** [comment]
I think trying different noise levels for different two-qubit gates is too expensive. 5**(number of two-qubit gates) is huge.
We can try the same noise level for all two-qubit gates in the circuit.
So if we have 5 noise levels, we have 5 error mitigation strategies.

**Comment by YuNariai** [comment]
when I test my code, I have a problem: I test with `execute` from tutorial example https://mitiq.readthedocs.io/en/stable/guide/calibrators.html
```
def execute(circuit, noise_level=0.001):
    circuit = circuit.with_noise(cirq.amplitude_damp(noise_level))
    result = cirq.DensityMatrixSimulator().run(circuit, repetitions=100)
    bitstrings = np.column_stack(list(result.measurements.values()))
    return MeasurementResult(bitstrings)
```
And the circuit from benchmarks:
```
circuit = generate_rb_circuits(2, 1)[0]
circuit.append(cirq.measure(circuit.all_qubits()))
```
And the reps use `represent_operation_with_local_depolarizing_noise` on two-qubit gate only.
I run `pec.execute_with_pec(circuit, execute, representations=reps)`, but get the error "Circuit has no measurements to sample."
Since I already append measure to all qubits, this warning confuses me

**Comment by andreamari** [comment]
Sounds like a bug. I just opened an issue #1805.

We never noticed this problem since we usually append measurements in the body of the executor.

In practice, you can do two things:

- Wait until #1805 is fixed.
- Try, as a workaround, to append measurements in the body of the executor. Not sure if this is easy/possible with the current structure of the `Calibrator` class.

**Comment by YuNariai** [comment]
for the execute I used, it also returns `MeasurementResult`,  doesn't it append measurement in the body of the executor?

**Comment by andreamari** [comment]
> for the execute I used, it also returns MeasurementResult, doesn't it append measurement in the body of the executor?


```
def execute(circuit, noise_level=0.001):
    circuit = circuit.with_noise(cirq.amplitude_damp(noise_level))
    circuit.append(cirq.measure(circuit.all_qubits()))   # <===========NEW LINE ADDED
    result = cirq.DensityMatrixSimulator().run(circuit, repetitions=100)
    bitstrings = np.column_stack(list(result.measurements.values()))
    return MeasurementResult(bitstrings)
```

**Comment by YuNariai** [comment]
with the new execute, I get a new warning
![image](https://user-images.githubusercontent.com/98194200/235266117-6143a4e1-8ff4-4952-bf9f-f13af445c901.png)
Is it because pec can not handle execute that returns measurements?

**Comment by andreamari** [comment]
Yes, `execute_with_pec`, expects an `execute` function that returns an expectation value (a `float`). 
The `execute` function can return a `MeasurementResult` only if the argument `observable` is not `None`.

Note however that the `Calibrator` instead requires an `execute` function which returns a `MeasurementResult` and it internally builds a new `expval_executor` from it, which returns an expectation value (a `float`) . 

This workflow is not specific to PEC. The same is true for `execute_with_zne` as well and similar functions.

**Comment by YuNariai** [comment]
I tried to append measurements in the body of the executor, but it's not supported by Calibrator. Here is the error I got
![image](https://user-images.githubusercontent.com/98194200/236362135-ad4851fa-013a-43af-bb28-61e5f329a3c2.png)
I guess I'll wait for PEC bug being solved

**Comment by Misty-W** [comment]
Hi @YuNariai, wanted to check if you're still working on this issue or if it's ok if someone else picks it up.

**Comment by YuNariai** [comment]
Hi @Misty-W , I just noticed #1805 was solved! So I can continue this now!

**Comment by Misty-W** [comment]
hi @YuNariai, thanks for your response! As it turns out, we already have a PR open that will probably be merged in the next couple days and solve this issue. 

However, that PR is just a basic prototype and we'll likely want to improve it. If you're interested in making some enhancements to the PEC calibration docs or workflow, just let us know!

**Comment by YuNariai** [comment]
OK! Go ahead to merge #1845 first! I'll check that to see if I can make any enhancements

**Comment by Misty-W** [comment]
> OK! Go ahead to merge #1845 first! I'll check that to see if I can make any enhancements. Btw in my code, errors will be applied on all two-qubit-gate, not only CNOT and CZ, do you think that's useful？

Awesome, thanks @YuNariai! 
Generalizing to representing all two-qubit gates in the circuit would be a good starting point.

**Comment by YuNariai** [comment]
OK! should I start a new pr after #1845 is merged?

**Comment by natestemen** [comment]
@YuNariai you could rebase your changes onto Misty's branch, and open a PR onto her branch. This might mean you will have to resolve some merge conflicts since #1845 is still in flux, but it might allow you to get some feedback sooner! Whatever works best for you, and thanks for your help.

**Comment by Misty-W** [comment]
Thanks @YuNariai! #1845 is merged now, so you can base your changes on `master`. 🎉
